docs/enhancements/0003-module-nmc-responsibilty-refactoring.md

```diff
@@ -0,0 +1,110 @@
+# Worker pods for KMM
+
+Authors: @yevgeny-shnaidman, @ybettan
+
+## Introduction
+
+This enhancement aims at redifining areas of resposiblity between Module-NMC and NMC controllers.
+This will allow for more clear-cut code and eliminate the variuos race-conditions that we are seeing(or will see) in the current situation
+
+### Current situation
+
+Currently both Module-NMC and NMC controller takes decision regarding kernel module deployment based on node status
+- Module-NMC controller check the schedulability of the node in order to decide whether kernel module should be deployed or removed
+  from the node (add/updating spec of the NMC or removing spec of the NMC)
+- NMC controller check the node's schedulability to decide whether to start creating loading/unloading pod on the node. In addition it also
+  check if the node has been recently rebooted, in order to create a loading pod, even if the status and spec of the NMC are equal.
+
+This creates a situation where 2 entities decide whether kernel modules should be loaded or not based on a nodes' status
+
+## Goals
+
+1. Create a clear-cut distinction between responsibilities of the two controllers
+2. Eliminate race conditions which are the result of the current situation
+
+## Non-Goals
+
+Do not change any other functionality of the two operators, besides their decision making that is described above
+
+## Design
+
+### Module-NMC controller decision-making flow 
+
+The flow takes into account both Module with Version field defined (ordered upgrade) and without Version field defined (un-ordered upgrade)
+Module-NMC does not take into account the current state of the Node (Ready/NotReady/Schedulable/etc'). It just defines if the kernel module should
+be loaded on the node or not based on whether there is a KernelMapping for the current node's kernel and on the labels of the node. All the rest of the decisions
+will be taken by NMC reconciler, which has a much better view of Node's current state and kernel module's current state
+
+1. Found out all the nodes targeted by the Module regardless of node's status, based on the node selector field of the Module
+2. If no suitable KernelMapping for the Node's kernel - do nothing
+3. If there is a suitable KernelMapping and Version field missing in Module (not an ordered upgrade) - update the spec
```

Comment:
```suggestion
3. If there is a suitable KernelMapping and Version field missing in Module (not an ordered upgrade) - update NMC's spec
```

Reply:
fixed

Reply:
I still see the old version. Did you push?